explorer/ast/statement.cpp

```diff
@@ -15,39 +15,86 @@ using llvm::cast;
 
 Statement::~Statement() = default;
 
-void Statement::PrintDepth(int depth, llvm::raw_ostream& out) const {
-  if (depth == 0) {
-    out << " ... ";
-    return;
+void Statement::PrintID(llvm::raw_ostream& out) const {
```

Comment:
While this is replacing the PrintDepth(1, ...) call, will this work as intended? Every use of `Print`, including the implicit ones like `<< match.expression() <<`, corresponds to a fully recursive call.

Maybe it should just print a keyword or brief placeholder matching the statement? e.g., `match` for Match, `{ ... }` for Block, PrintID for ExpressionStatement, etc. This would be similar to how expressions print IDs.

Reply:
Oh I'm sorry I missed that. I'll do that in the next commit. Thanks!

Reply:
I think that PrintID should not do anything that is recursive or loops. If you are fine with recursion or looping, you should probably be using Print instead.

Reply:
Oh, to get rid of this looping issue with the block can we have the following cases?

if the number of statements inside the block are less than or equal to 2
```
{id stmt 1; id stmt 2}
```
if the statements in the block are greater than 2
```
{id stmt 1; ... id stmt2}
```

although in the trace output, source location is also added in the cases when `PrintID` is used but printing `{...}` for blocks is making it a bit hard to understand which part of code it belongs to.

Reply:
That would be fine, but along similar lines you might consider:

0: `{}`
1: `{id stmt 1;}`
N: `{id stmt 1; ...[N-1] statements...}`

i.e., after the 1st statement, just print the # of statements that were elided.

Reply:
Nice suggestion, I will implement that.

Reply:
I also used `PrintID` at places where I felt like the information could be useful. Let me know what you think about it.
And I noticed an issue in expression's `PrintID` due to it there was an assertion error happening, I also tried to fix that.

Reply:
> I'm still concerned about this: while a Block statement would just be { ... } which could be confusing, there are more cues (keywords) for other pieces. I think too that not recursing is more in keeping with the prior PrintDepth implementation.

Oh okay. Thanks for letting me know about that!

> Note especially that your changes to For are back to recursively printing the entire for loop contents. This is returning to the issue I was trying to point out originally, that PrintID needs to not be fully recursive.

I actually forgot to use `PrintAsID(...)` there but going back to simple implementation.

> I would appreciate going back to a simple implementation so that we can finish review.

I'm reverting back to the simple implementation in next commit.

Reply:
I do understand that non-recursive yields less information, but I think things like `{var x: T = {}.(interface ImplicitAs(T = class T).Convert)();return 0;}` (which had been in trace output before) are actually more indicative of an issue that needed to be fixed, that arbitrary amounts of output were getting printed.

I think this current setup is good, thanks!